doc(version) Some initial work on documentation for 0.21

jeromesimeon opened the pull request:
Signed-off-by: Jerome Simeon <jeromesimeon@me.com>

### #306 

- Some simplification of the documentation Table of Content: Guides first, and Tutorials consolidated under a single Section
- Moved around some of the material in Introduction
   - "lawyer/business/developer" now in main overview
   - Distinguishes intro material on "smart legal contract" v "templates"
   - Move online tour to Introduction
   - better outline for the documentation
- Full revision to "markdown text" guide with CiceroMark & TemplateMark for Cicero 0.21
   - New CiceroMark section for contract text
   - Fully revised TemplateMark Section with new documentation for optional blocks and join blocks
   - Distinguishes "inline blocks" from "container blocks" following commonmark spec
- Full review / fixes to tutorials
   - Tested with cicero 0.21

dselman commented on log/trace.log:
```diff
@@ -0,0 +1,116 @@
+2:50:41 PM - [32minfo[39m: Loading a default text/sample.md file.
```
Do we need to add this?

dselman commented:
Looks great. A few nit picks.

1. The `ulist` and `olist` examples do not obviously follow the layout for a container block described above, which people may find confusing:
https://deploy-preview-309--accordproject-docs.netlify.app/docs/next/markup-templatemark.html#unordered-lists

2. I think it would be good to point out that the contents of formulas are not parsed: https://deploy-preview-309--accordproject-docs.netlify.app/docs/next/markup-templatemark.html#ergo-formulas

So, they allow the template developer to generate arbitrary (soon!) markdown text from other contract and clause variables. They therefore cannot be used to *set* a template model variable during parsing.

dselman reviewed (changes requested):
Looks amazing. Couple of nit picks - also looks like a log file was accidentally added.

The `ulist` and `olist` examples do not obviously follow the layout for a container block described above, which people may find confusing:
https://deploy-preview-309--accordproject-docs.netlify.app/docs/next/markup-templatemark.html#unordered-lists

I think it would be good to point out that the contents of formulas are not parsed: https://deploy-preview-309--accordproject-docs.netlify.app/docs/next/markup-templatemark.html#ergo-formulas

Formulas allow the template developer to generate arbitrary (soon!) markdown text from other contract and clause variables. They therefore cannot be used to set a template model variable during parsing.

dselman commented:
Once this is merged I will add a reference for the new decorators. 

I'd love to also be cheeky and ask for a snippet of Ergo that shows how to process a request and conditionally emit an obligation .... I can never remember how to do that! :-) 

Perhaps we could add a sample clause to this to summarise how to bringdifferent capabilities together: processing a request to return a response, using an enforce, printing and emitting a conditional obligation?

https://deploy-preview-309--accordproject-docs.netlify.app/docs/logic-stmt.html#emitting-obligations

mttrbrts commented:
It looks like there is a lot of redundant stuff that has been copy/pasted from the 0.13->0.20 migration guide to the 0.20->0.21 guide. Is there another doc somewhere?

Also, presumable we should create a new Docusaurus release to make this available by default?

peterhunn reviewed (approved):
LGTM

jeromesimeon commented:
> Looks amazing. Couple of nit picks - also looks like a log file was accidentally added.
> 

Fixed.

> The `ulist` and `olist` examples do not obviously follow the layout for a container block described above, which people may find confusing:
> https://deploy-preview-309--accordproject-docs.netlify.app/docs/next/markup-templatemark.html#unordered-lists
> 

Good catch! Would indeed be really confusing. Fixed.

> I think it would be good to point out that the contents of formulas are not parsed: https://deploy-preview-309--accordproject-docs.netlify.app/docs/next/markup-templatemark.html#ergo-formulas
> 
> Formulas allow the template developer to generate arbitrary (soon!) markdown text from other contract and clause variables. They therefore cannot be used to set a template model variable during parsing.

Fixed. Thanks for the suggested text.

jeromesimeon commented:
> It looks like there is a lot of redundant stuff that has been copy/pasted from the 0.13->0.20 migration guide to the 0.20->0.21 guide. Is there another doc somewhere?
> 

Thanks. Yes I was still in the middle of it. Should look better now.

> Also, presumable we should create a new Docusaurus release to make this available by default?

Agreed! I'll merge after a second round of review and then tag.

jeromesimeon commented:
@dselman @mttrbrts All comments should now have been addressed.

jeromesimeon commented:
> Once this is merged I will add a reference for the new decorators.
> 
> I'd love to also be cheeky and ask for a snippet of Ergo that shows how to process a request and conditionally emit an obligation .... I can never remember how to do that! :-)
> 
> Perhaps we could add a sample clause to this to summarise how to bring different capabilities together: processing a request to return a response, using an enforce, printing, and emitting a conditional obligation?
> 
> https://deploy-preview-309--accordproject-docs.netlify.app/docs/logic-stmt.html#emitting-obligations



> Once this is merged I will add a reference for the new decorators.
> 
> I'd love to also be cheeky and ask for a snippet of Ergo that shows how to process a request and conditionally emit an obligation .... I can never remember how to do that! :-)
> 
> Perhaps we could add a sample clause to this to summarise how to bring different capabilities together: processing a request to return a response, using an enforce, printing, and emitting a conditional obligation?
> 
> https://deploy-preview-309--accordproject-docs.netlify.app/docs/logic-stmt.html#emitting-obligations

Opened a new issue to track things like this: https://github.com/accordproject/techdocs/issues/310
